x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/CancelJobModelSnapshotUpgradeAction.java

```diff
@@ -0,0 +1,200 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.ml.action;
+
+import org.elasticsearch.action.ActionRequestValidationException;
+import org.elasticsearch.action.ActionType;
+import org.elasticsearch.action.support.tasks.BaseTasksRequest;
+import org.elasticsearch.action.support.tasks.BaseTasksResponse;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.core.TimeValue;
+import org.elasticsearch.xcontent.ObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xpack.core.ml.job.config.Job;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+
+import java.io.IOException;
+import java.util.Objects;
+import java.util.concurrent.TimeUnit;
+
+public class CancelJobModelSnapshotUpgradeAction extends ActionType<CancelJobModelSnapshotUpgradeAction.Response> {
+
+    public static final CancelJobModelSnapshotUpgradeAction INSTANCE = new CancelJobModelSnapshotUpgradeAction();
+    public static final String NAME = "cluster:admin/xpack/ml/job/model_snapshots/upgrade/cancel";
```

Comment:
```suggestion
    public static final String NAME = "cluster:internal/xpack/ml/job/model_snapshots/upgrade/cancel";
```

Since there is no rest layer for this, it seems prudent to be internal.

Reply:
I intentionally made this `admin` instead of `internal`.  Although there's no REST layer today it's possible that we might need to add one in the future, say if we ever want the UI to be able to clean up a failed upgrade.

The reason we've used `internal` instead of `admin` in the past is to make it hard for users of the transport client to call the API, even though it's easy for them to write the code to call it.  Since there's nothing inherently wrong with calling this endpoint if you want to clean up failed upgrades I don't think that's a problem in this case.  (Plus it's only in 7.17 that the transport client will even exist - after that lack of a REST layer will make this impossible to call.)  I will add a comment to say this.

Part of the reason I didn't add a REST layer in this PR was because it's being added so late in the 7.x development cycle and it's not strictly essential at the moment.

Reply:
`internal` conveys a message even if just a convention. If this was to become a public API with a REST interface we would create a new `admin` wrapper action that calls this one and have this action `internal`. We have many internal APIs called this way. 

I don't consider this a blocker on the PR but there is a path for making `internal` APIs public, if you think it likely there will be a REST API then keep it `admin` otherwise it makes more sense to be `internal`.

Reply:
I think it's 50/50 whether this will be made public in the future.

> If this was to become a public API with a REST interface we would create a new `admin` wrapper action that calls this one and have this action `internal`.

This seems like needless baggage to me.  The wrapper action would literally call the internal action and do nothing else.  It would just be extra boilerplate code.

> `internal` conveys a message even if just a convention.

It's only the ML team that has ever done this.  All the internal actions of core Elasticsearch start with either `admin` or `monitor`.  (For example look at the persistent tasks actions.)  The original reason why the ML team did this was to make transport client users think twice before calling an API that was intended to be internal.  Of course this only had an effect if security was enabled, but it was in the days of X-Pack as a separate plugin where ML users were likely to be using security but most users weren't using security.  It was also in the days before the transport client was deprecated so we were more worried about transport client users accidentally automating things they shouldn't.

Be as I said in the comment I added to the code, in this case we wouldn't really care if a transport client user did call this action.  It doesn't do something low level that will mess up the cluster if called at the wrong time.  It's an action that _could_ be external, just isn't at the moment.

Reply:
I agree that as the transport client is not used directly any more we don't really gain much by calling the action `internal`.

Reply:
Its good we had this discussion then as `internal` is a convention we have been using for a long time. If we chose to stop doing this then we have already established the usage of naming internal actions internal e.g. `InternalInferModelAction` perhaps we can switch to that convention for actions that are purely internal.

Reply:
The key difference with this action is that it doesn't _need_ to be internal.  It should be possible to make it external in the future with as little extra effort as possible.  Whereas `InternalInferModelAction` is specifically designed to be internal, and it would cause problems if people ever started calling it externally.  An action that's more similar to this one is `DeleteExpiredDataAction`, where its primary purpose is to be called by the daily maintenance task, but it doesn't hurt to call it externally.  `DeleteExpiredDataAction` does have a corresponding REST endpoint, but if we'd wanted to avoid that initially to avoid yet another endpoint that shows up in our list of documented endpoints then we could have deferred adding the REST endpoint to a release after the action was added.

Reply:
> It's only the ML team that has ever done this. All the internal actions of core Elasticsearch start with either admin or monitor.

> I agree that as the transport client is not used directly any more we don't really gain much by calling the action internal.

I was replying to these points stating that the original reason for using `internal` no longer applies and that and suggesting it can be discontinued with the demise of the transport client.

Reply:
> I was replying to these points stating that the original reason for using `internal` no longer applies and that and suggesting it can be discontinued with the demise of the transport client.

Yes, sure, for new actions we can discontinue it.  But for the existing ones that contain `internal` changing it is a non-trivial piece of work, so we might as well leave them as-is.